src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java

```diff
@@ -15,48 +13,391 @@
+        System.arraycopy(command, 0, cmd, 0, command.length);
+        for(int i=0; i<cmd.length; i++) {
+           if(cmd[i].indexOf(INPUT_FILE_TOKEN) != -1) {
+              cmd[i] = cmd[i].replace(INPUT_FILE_TOKEN, stream.getFile().getPath());
+              inputToStdIn = false;
+           }
+           if(cmd[i].indexOf(OUTPUT_FILE_TOKEN) != -1) {
+              output = tmp.createTemporaryFile();
+              outputFromStdOut = false;
+           }
         }
+
+        // Execute
+        Process process;
+        if(cmd.length == 1) {
+           process = Runtime.getRuntime().exec( cmd[0] );
+        } else {
+           process = Runtime.getRuntime().exec( cmd );
+        }
+
+        Parser stdoutParser = null;
+        Parser stderrParser = null;
+        try {
+            if(inputToStdIn) {
+               sendInput(process, stream);
+            } else {
+               process.getOutputStream().close();
+            }
+
+            InputStream out = process.getInputStream();
+            InputStream err = process.getErrorStream();
+            
+            if(hasPatterns) {
+               stderrParser = extractMetadata(err, metadata);
+               
+               if(outputFromStdOut) {
+                  extractOutput(out, xhtml);
+               } else {
+                  stdoutParser = extractMetadata(out, metadata);
+               }
+            } else {
+               ignoreStream(err);
+               
+               if(outputFromStdOut) {
+                  extractOutput(out, xhtml);
+               } else {
+                  ignoreStream(out);
+               }
+            }
+        } finally {
+            try {
+                process.waitFor();
+            } catch (InterruptedException ignore) {
+            }
+        }
+
+        // Grab the output if we haven't already
+        if (!outputFromStdOut) {
+            extractOutput(new FileInputStream(output), xhtml);
+        }
+        
+        waitForParserToFinish(stdoutParser);
+        waitForParserToFinish(stderrParser);
+    }
+    
+    private void waitForParserToFinish(Parser parser) throws IOException
+    {
+        if (parser != null)
+        {
+            try 
+            {
+                parser.join(MAX_WAIT_TIME_MS);
+            }
+            catch (InterruptedException ignore)
+            {
+                // This happens only if client code calls interrupt on this thread.
+            }
+            
+            // Checks that the parser finished successfully
+            if (parser.isDone)
+            {
+                if (parser.ioException != null)
+                {
+                    // The parser was stopped by an IOException, so let the client know.
+                    throw parser.ioException;
+                }
+            }
+            else
+            {
+                parser.interrupt();
+                throw new IOException("The parser took too long to parse the output from the external command.");
```

Comment:
This should be a `TimeoutException` and perhaps mention the timeout value: "... took longer than XXXms".  The `parse` method can catch that and rethrow as a `TikaException` if need be.